File: src/Leitor/LeitorEstudante.java
package Leitor;

import java.util.Scanner;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.io.Serializable;
import java.util.HashMap;

import Modelo.Estudante;

public class LeitorEstudante extends ILeitor implements Serializable{
    
    private static final long serialVersionUID = 1L;
    private HashMap<String, Estudante> mapa = new HashMap<>();
    private static LeitorEstudante leitor;

    private LeitorEstudante(){}

    public static LeitorEstudante obterInstancia(){
        if (leitor == null){
            leitor = new LeitorEstudante();
            return leitor;
        }
        else return leitor;
    }

    public Set<String> obterChaves(){
        return mapa.keySet(); 
    }

    public HashMap<String,Estudante> obterHash(){
        return mapa;
    }

    public static void setInstancia(LeitorEstudante novo){
        leitor = novo;
    }
    
    public Estudante ler(Scanner scanner){
        String matricula;
        
        matricula = scanner.next().trim();
        
        String pattern = "[^0-9]";
        Pattern r = Pattern.compile(pattern);
        Matcher m = r.matcher(matricula);

        if(m.find()){
            throw new RuntimeException("Dado inválido: "+matricula+".");
        }
        
        if(busca(matricula) != null)
			throw new IllegalArgumentException("Cadastro repetido: "+matricula+".");
        String nome = scanner.next().trim().replaceAll(" +", " ");
        Estudante aluno = new Estudante(matricula, nome);
        if(scanner.hasNext())
            scanner.nextLine();
        return aluno;
    }

    public void anexaHash(Estudante aluno){
        mapa.put(aluno.obterMatricula(), aluno);
    }

    public Estudante busca(String matricula){
        if(mapa.containsKey(matricula)){
            return mapa.get(matricula);
        }
        return null;
    }
}

File: cpp_version/Leitor/LeitorEstudante.cpp
#include "LeitorEstudante.h"

LeitorEstudante* LeitorEstudante::leitor = NULL;

LeitorEstudante::LeitorEstudante(){}

LeitorEstudante* LeitorEstudante::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorEstudante();
        return leitor;
    }
    else return leitor;
}

vector<Estudante*> LeitorEstudante::obterEstudantes(){
    vector<Estudante*> lista;
    for(map<string,Estudante*>::iterator it = mapEstudante.begin(); it != mapEstudante.end(); ++it){
        lista.push_back(it->second);
    }
    return lista;
}

Estudante* LeitorEstudante::ler(vector<string> vec){
    string matricula = vec[0];

    regex rgx("[^0-9]");
    smatch match;
    if(regex_search(matricula, match, rgx))
        throw ExcecaoDado("Dado inválido: "+matricula+".");

    if(busca(matricula) != NULL)
        throw ExcecaoCad("Cadastro repetido: "+matricula+".");
    
    string aux = vec[1];

    regex rgx2(" +");
    string nome = regex_replace(aux, rgx2, " ");

    return new Estudante(matricula, nome);
}

void LeitorEstudante::anexaHash(Estudante* estudante){
    mapEstudante[estudante->obterMatricula()] = estudante;
}

Estudante* LeitorEstudante::busca(string matricula){
    bool contains = false;
    for(map<string,Estudante*>::iterator it = mapEstudante.begin(); it != mapEstudante.end(); ++it){
        if(it->first == matricula){
            contains = true;
        }
    }
    if(contains){
        return mapEstudante[matricula];
    }
    return NULL;
}